Mathlib/Data/List/Sym.lean

```diff
@@ -0,0 +1,264 @@
+/-
+Copyright (c) 2023 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+import Mathlib.Data.Nat.Choose.Basic
+import Mathlib.Data.Sym.Sym2
+
+/-! # Unordered tuples of elements of a list
+
+Defines `List.sym` and the specialized `List.sym2` for computing lists of all unordered n-tuples
+from a given list. These are list versions of `Nat.multichoose`.
+
+## Main declarations
+
+* `List.sym`: `xs.sym n` is a list of all unordered n-tuples of elements from `xs`,
+  with multiplicity. The list's values are in `Sym α n`.
+* `List.sym2`: `xs.sym2` is a list of all unordered pairs of elements from `xs`,
+  with multiplicity. The list's values are in `Sym2 α`.
+
+## Todo
+
+* Prove `protected theorem Perm.sym (n : ℕ) {xs ys : List α} (h : xs ~ ys) : xs.sym n ~ ys.sym n`
+  and lift the result to `Multiset` and `Finset`.
+
+-/
+
+namespace List
+
+variable {α : Type*}
+
+section Sym2
+
+/-- `xs.sym2` is a list of all unordered pairs of elements from `xs`.
+If `xs` has no duplicates then neither does `xs.sym2`. -/
+protected def sym2 : List α → List (Sym2 α)
+  | [] => []
+  | x::xs => ((x::xs).map fun y => ⟦(x, y)⟧) ++ xs.sym2
```

Comment:
I don't really know what's idiomatic nowadays but I was expecting something like
```suggestion
  | x :: xs => (x :: xs).map (⟦(x, ·)⟧) ++ xs.sym2
```

Reply:
Using `·` didn't work, but I put the spaces around `::`.

Reply:
Try `(⟦⟨x, ·⟩⟧)`. I think the innermost pair of parentheses will be interpreted as a function notation.

Reply:
I don't think this is worth changing, but your suggestion probably works (modulo questions about whether it can infer that it should be a Prod).